### ISSUE TITLE
erroneous "unnecessary use of `to_string`"

### ISSUE THREAD

**Issue by hexmode** [rust-lang/rust-clippy#10033]
### Summary

Clippy complains about an extra to_string when a `Deref` implementation is provided that results in something different result than `to_string`. I originally described this problem at [Why does clippy want two to_string() calls in a row? ](https://users.rust-lang.org/t/why-does-clippy-want-two-to-string-calls-in-a-row/85381) and @zirconium-n provided the example case here.

### Lint Name

clippy::unnecessary_to_owned

### Reproducer

I tried this code:

```rust
use std::{fmt::Display, ops::Deref};

fn main() {
    let f = Foo;

    foo(&f.to_string())
}

fn foo(s: &str) {
    println!("{}", s);
}

struct Foo;

impl Deref for Foo {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        "str"
    }
}

impl Display for Foo {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Foo")
    }
}
```

I saw this happen:

```
warning: unnecessary use of `to_string`
 --> src/main.rs:6:11
  |
6 |     foo(&f.to_string())
  |           ^^^^^^^^^^^^ help: remove this
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
  = note: `#[warn(clippy::unnecessary_to_owned)]` on by default

warning: `clippy-bug` (bin "clippy-bug") generated 1 warning
```

I expected to see this happen: (no errors)

### Version

```text
rustc 1.67.0-nightly (09508489e 2022-11-04)
binary: rustc
commit-hash: 09508489efc223287731fe8abbd2a81bbf7adf8e
commit-date: 2022-11-04
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
